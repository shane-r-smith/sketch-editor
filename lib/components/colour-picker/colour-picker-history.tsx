import type { FunctionComponent } from "react";
import type { ColourPickerHistoryProps } from "./colour-picker.types";

import Grid from "@mui/material/Grid";
import Box from "@mui/material/Box";

export const ColourPickerHistory: FunctionComponent<
  ColourPickerHistoryProps
> = ({ history, numHistoryToShow, onColourSelected }) => {
  const unique = Array.from(new Set(history));
  const historyItemsToShow =
    unique.length < numHistoryToShow
      ? [...unique, ...Array(numHistoryToShow - unique.length).fill(undefined)]
      : unique.slice(0, numHistoryToShow);

  return (
    <Grid container spacing={1}>
      {historyItemsToShow.map((colour, i) => {
        return (
          <Box
            key={`${colour}-colour-history-${i}`}
            sx={[
              {
                width: "1.75rem",
                height: "1.75rem",
                borderRadius: "0.25rem",
                backgroundColor: "grey.50",
              },
              colour && {
                cursor: "pointer",
                backgroundColor: colour,
                "&:hover": {
                  outlineWidth: 2,
                  outlineStyle: "solid",
                  outlineColor: "primary.main",
                },
              },
            ]}
            onClick={() => {
              if (!colour) {
                return;
              }

              onColourSelected(colour);
            }}
          />
        );
      })}
    </Grid>
  );
};

export default ColourPickerHistory;
