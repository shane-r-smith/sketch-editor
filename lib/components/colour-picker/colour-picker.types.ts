import type { ColourType } from "../../domain";

export interface BaseColourPickerProps {
  /**
   * The initial colour value for the colour picker.
   * This is the colour that will be displayed when the picker is first rendered.
   */
  colour: string;

  /**
   * Type of colour input to use.
   * This can be "HEX" for hexadecimal input or "RGB" for RGB input.
   */
  type: ColourType;

  /**
   * Callback function to handle colour changes.
   * This function is called when the user selects a colour.
   * @param colour The selected colour value.
   * @param type The type of colour (e.g., HEX, RGB).
   */
  onChange: (colour: string, type: ColourType) => void;
}

export interface ColourPickerHistoryProps {
  /**
   * History of colours used in the picker.
   * This is used to show the last used colours
   * and allow users to quickly select them.
   */
  history: string[];

  /**
   * Number of history items to show in the picker.
   * This is used to limit the number of items shown.
   */
  numHistoryToShow: number;

  onColourSelected: (colour: string) => void;
}

export interface ColourPickerProps
  extends BaseColourPickerProps,
    ColourPickerHistoryProps {}
