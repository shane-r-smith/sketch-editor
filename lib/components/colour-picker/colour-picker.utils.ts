type Enumerate<
  N extends number,
  Acc extends number[] = []
> = Acc["length"] extends N
  ? Acc[number]
  : Enumerate<N, [...Acc, Acc["length"]]>;

type IntRange<F extends number, T extends number> = Exclude<
  Enumerate<T>,
  Enumerate<F>
>;

export const stringToClampedRgb = (digitString: string): IntRange<0, 256> => {
  const _digit = parseInt(digitString);

  if (isNaN(_digit)) {
    return 0;
  }

  return clampRgbDigit(_digit);
};

/**
 * Ensures that a given number is a valid RGB digit.
 * If the input is less than 0, returns 0.
 * If the input is greater than 255, returns 255.
 * Otherwise, returns the input unchanged.
 *
 * @param digit - The number to validate as an RGB digit.
 * @returns The validated RGB digit, clamped between 0 and 255.
 */
export const clampRgbDigit = (digit: number): IntRange<0, 256> => {
  if (digit < 0) {
    return 0;
  }

  if (digit > 255) {
    return 255;
  }

  return parseInt(`${digit}`) as IntRange<0, 256>;
};
