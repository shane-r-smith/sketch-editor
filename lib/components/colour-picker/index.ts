export * from "./colour-picker";
export * from "./colour-picker.types";
export * from "./colour-picker-history";
export * from "./colour-picker.utils";

export * from "./colour-picker-type/colour-picker-type";
export * from "./colour-picker-type/typed-colour-picker";

export * from "./colour-inputs/colour-input";
export * from "./colour-inputs/hex-colour-input";
export * from "./colour-inputs/rgb-colour-input";
