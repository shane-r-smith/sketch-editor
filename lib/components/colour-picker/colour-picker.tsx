import { type ColourPickerProps } from "./colour-picker.types";
import { TypedColourPicker } from "./colour-picker-type/typed-colour-picker";
import { ColourPickerType } from "./colour-picker-type/colour-picker-type";
import { ColourInput } from "./colour-inputs/colour-input";
import ColourPickerHistory from "./colour-picker-history";
import Grid from "@mui/material/Grid";

export function ColourPicker({
  colour,
  type,
  history,
  numHistoryToShow,
  onColourSelected,
  onChange,
}: ColourPickerProps) {
  return (
    <Grid container direction="column" spacing={2}>
      <TypedColourPicker colour={colour} type={type} onChange={onChange} />
      <ColourPickerType colour={colour} type={type} onChange={onChange} />
      <ColourInput colour={colour} type={type} onChange={onChange} />
      <ColourPickerHistory
        history={history}
        numHistoryToShow={numHistoryToShow}
        onColourSelected={onColourSelected}
      />
    </Grid>
  );
}
