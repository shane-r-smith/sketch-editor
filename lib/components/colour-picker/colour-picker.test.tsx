import { render } from "@testing-library/react";
import { ColourPicker } from "./colour-picker";
import type { ColourPickerProps } from "./colour-picker.types";
import userEvent from "@testing-library/user-event";

describe("<ColourPicker />", () => {
  const defaultProps: ColourPickerProps = {
    colour: "#ff0000",
    type: "HEX",
    history: ["#ff0000", "#00ff00", "#0000ff"],
    numHistoryToShow: 3,
    onColourSelected: vi.fn(),
    onChange: vi.fn(),
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<ColourPicker {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("calls onChange when colour input changes", async () => {
    // Arrange
    const user = userEvent.setup();

    // Act
    const { getByDisplayValue } = render(<ColourPicker {...defaultProps} />);
    const input = getByDisplayValue(defaultProps.colour);
    await user.clear(input);
    await user.type(input, "#123456");

    // Assert
    expect(defaultProps.onChange).toHaveBeenCalled();
  });
});
