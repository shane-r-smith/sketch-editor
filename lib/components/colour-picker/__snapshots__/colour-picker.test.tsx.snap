// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ColourPicker /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-spacing-xs-2 css-1cax9kl-MuiGrid-root"
  >
    <div
      class="react-colorful _editorColourPicker_cffd98"
    >
      <div
        class="react-colorful__saturation"
        style="background-color: rgb(255, 0, 0);"
      >
        <div
          aria-label="Color"
          aria-valuetext="Saturation 100%, Brightness 100%"
          class="react-colorful__interactive"
          role="slider"
          tabindex="0"
        >
          <div
            class="react-colorful__pointer react-colorful__saturation-pointer"
            style="top: 0%; left: 100%;"
          >
            <div
              class="react-colorful__pointer-fill"
              style="background-color: rgb(255, 0, 0);"
            />
          </div>
        </div>
      </div>
      <div
        class="react-colorful__hue react-colorful__last-control"
      >
        <div
          aria-label="Hue"
          aria-valuemax="360"
          aria-valuemin="0"
          aria-valuenow="0"
          class="react-colorful__interactive"
          role="slider"
          tabindex="0"
        >
          <div
            class="react-colorful__pointer react-colorful__hue-pointer"
            style="top: 50%; left: 0%;"
          >
            <div
              class="react-colorful__pointer-fill"
              style="background-color: rgb(255, 0, 0);"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiSelect-root css-sc8y68-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-trm6af-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          role="combobox"
          tabindex="0"
        >
          HEX
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
          tabindex="-1"
          value="HEX"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1nf2c5d-MuiNotchedOutlined-root"
          >
            <span
              aria-hidden="true"
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiStack-root css-nen11g-MuiStack-root"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-colorPrimary css-nwsdb5-MuiFormLabel-root"
      >
        HEX
      </label>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r1»"
            type="text"
            value="#ff0000"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1nf2c5d-MuiNotchedOutlined-root"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-1 css-1yvsbrk-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-10zmmb7"
      />
      <div
        class="MuiBox-root css-1wy9ye4"
      />
      <div
        class="MuiBox-root css-1qu7dg4"
      />
    </div>
  </div>
</div>
`;
